### ISSUE TITLE
Installer ComponentAdapter 

### ISSUE THREAD

**Issue by JProof** [joomla/joomla-cms#36006]
### Steps to reproduce the issue

During component-development directly in joomla

* Remove from a installable component in the  `<administration> </administration>`
* Install the component by "discover"
* After discover the following screen
![img-2021-11-10-20-38-26](https://user-images.githubusercontent.com/12245927/141182181-a5d580ea-47fb-4e94-aef4-93a669636cc1.png)

### Expected result

Not installed should really not installed

### Actual result

The Component was installed/discovered into the Database
![img-2021-11-10-20-40-28](https://user-images.githubusercontent.com/12245927/141182319-3e1d485f-7e1d-4a5d-b29e-04827ba45a51.png)

The Problem is, that the component namespace does not work because component is HALF installed 

### System information (as much as possible)
Joomla! 4.0.4
every system 

### Additional comments

Breaks here:

![img-2021-11-10-20-46-48](https://user-images.githubusercontent.com/12245927/141183037-a0b9eb99-a712-46e4-8ce2-55b4f60bf3ab.png)

After Breaking final installation tasks will not finished. So the the component-namespace was not "registered" somewhere 

![img-2021-11-10-20-47-17](https://user-images.githubusercontent.com/12245927/141183065-fb28156a-7591-40b1-886d-03fdf9ab2b85.png)

EDIT:
In addition: is it possible or better for development during discover to re-create / handle the namespace-cache generation with the PlgExtensionNamespacemap???

**Comment by dgrammatiko** [comment]
This is a known issue due to some limitations from the J3 code: https://github.com/mbabker/com_empty/blob/4a199f19b2d7ab240b6092057b7edbff666818e9/com_empty/empty.xml#L18-L20

**Comment by alikon** [comment]
funny

> Sad panda

**Comment by dgrammatiko** [comment]
@alikon I think that the check for the administrator in the manifest should be done much earlier and bail out without doing anything more (now it copies files, etc). So, I guess this is still a valid issue

**Comment by JProof** [comment]
Okay!

By the way, the `<administrator><menu>...` also never works on discover, perhaps that is the right time to change this "ugly" behaviours?

**Comment by dgrammatiko** [comment]
> perhaps that is the right time to change this "ugly" behaviours?

You're welcome to contribute some code

**Comment by alikon** [comment]
yes for sure, @JProof please submit a pr


i've always thinked that
`<administration>
	</administration>`

was mandatory

anyway happy to be wrong :smile:

**Comment by dgrammatiko** [comment]
> i've always thinked that `<administration> </administration>` was mandatory

That's right, it is **mandatory**. That said if it's missing it shouldn't produce half installed components

There's a PR that changes the order of execution so the installer will bail out without writting any files: https://github.com/joomla/joomla-cms/pull/36014

**Comment by JProof** [comment]
will look at the weekend for the following "issues"
* this issue
* the `<administration><menu` (f**ks me of since a couple of years xD) reinstall the admin component menu items(if changed)
* and the `<namespace` in manifest 

Goal is develop directly in joomla fast/simple components and "discover" that the component is working.

cu

**Comment by dgrammatiko** [comment]
@JProof you could test #36014. It doesn't solve all the issues but at least won't produce the current mess (I hope)

**Comment by JProof** [comment]
@dgrammatiko yes ..thanks. will look for it over the weekend.

**Comment by JProof** [comment]
@dgrammatiko as the `<administration>` is mandatory, why we throw the RuntimeException? 
That makes no sense for me.

```php

// Make sure that we have an admin element
		if (!$this->getManifest()->administration)
		{
			throw new \RuntimeException(Text::_('JLIB_INSTALLER_ERROR_COMP_INSTALL_ADMIN_ELEMENT'));
		}
```

**Comment by dgrammatiko** [comment]
> as the <administration> is mandatory, why we throw the RuntimeException?

I guess it was supposed to be a fail-safe switch but it ran too late, the conditional should be an indication if the installation should proceed or not. The trhowable it's more to do with the parent code (the try/catch part)

**Comment by JProof** [comment]
Okay, thank you!
last question for the moment i hope 

https://github.com/joomla/joomla-cms/tree/4.0-dev/tests/Unit/Libraries/Cms/Installer/Adapter
There is no unit.test for the ComponentAdapter? or i am in wrong location?

**Comment by dgrammatiko** [comment]
> There is no unit.test for the ComponentAdapter? or i am in wrong location?

You're right, no unit tests
